### PR TITLE
docs: Add installation and usage to README

### PR DESCRIPTION
**Pull request by kurtisvg**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,124 @@
 # cloud-sql-go-connector
 *Warning*: This project is experimental, and is not an officially supported 
 Google product.
+
+The _Cloud SQL Go Connector_ provides IAM authorization to an app's connections
```
Comment (enocom) [inline README.md:5]: authorization for

Comment (enocom) [reply README.md:5]: or better: provides IAM authorization and an encrypted connection for an app's connections to a Cloud SQL instance.

Comment (enocom) [reply README.md:5]: In other words, should encryption be in here as one of the selling points?

File: README.md [hunk 2]
```diff
@@ -1,3 +1,124 @@
 # cloud-sql-go-connector
 *Warning*: This project is experimental, and is not an officially supported 
 Google product.
+
+The _Cloud SQL Go Connector_ provides IAM authorization to an app's connections 
+to a Cloud SQL instance. It provides connection level authorization only; it 
+_does not_ provide a network path to an instance that doesn't already exist 
+(ie. you will still be unable to connect via Private IP without access to the
```
Comment (enocom) [inline README.md:8]: i.e.

Comment (enocom) [reply README.md:8]: connect via Private IP, or connect to a Private IP database? The word "via" seems strange to me here.

File: README.md [hunk 3]
```diff
@@ -1,3 +1,124 @@
...
+to a Cloud SQL instance. It provides connection level authorization only; it 
+_does not_ provide a network path to an instance that doesn't already exist 
+(ie. you will still be unable to connect via Private IP without access to the 
+correct VPC). For more information, see 
+[About the Cloud SQL Auth proxy][about-proxy].
+
+[about-proxy]: https://cloud.google.com/sql/docs/mysql/sql-proxy
+
+The _Cloud SQL Go Connector_ is an experimental new version of the 
+[Cloud SQL proxy dialer](). It's API is considered unstable and may change in
```
Comment (enocom) [inline README.md:15]: There's a missing link here.

Comment (enocom) [reply README.md:15]: Also "Its"

File: README.md [hunk 4]
```diff
@@ -1,3 +1,124 @@
...
+
+The _Cloud SQL Go Connector_ is an experimental new version of the 
+[Cloud SQL proxy dialer](). It's API is considered unstable and may change in 
+the future. Please use at your own risk. 
+
+[proxy-dialer]: https://github.com/GoogleCloudPlatform/cloudsql-proxy/tree/main/proxy#cloud-sql-proxy-dialer-for-go
+
+## Installation
+
+First, clone this repo into a folder relative to your project:
```
Comment (enocom) [inline README.md:22]: Should there be a short comment like, "Until this module is published, we will manually install it using a replace directive"?

File: README.md [hunk 5]
```diff
@@ -1,3 +1,124 @@
...
+
+Next, use a _replace_ directive in your `go.mod` that points to the cloned folder:
+```
+replace cloud.google.com/cloudsqlconn => ../cloud-sql-go-connector
+```
+
+## Usage 
+
+This package provides several functions for authorizing and encrypting 
+connections on your behalf. Typically, these functions will pass these
```
Comment (enocom) [inline README.md:35]: The sentence starting with "Typically..." is hard to parse. Could we remove it?

File: README.md [hunk 6]
```diff
@@ -1,3 +1,124 @@
...
+The instance connection name for your Cloud SQL instance is always in the 
+format "project:region:instance".
+
+### Credentials 
+
+This repo uses the [Application Default Credentials (ADC)][adc] strategy for 
+providing credentials. Please see the [golang.org/x/oauth2/google][google-auth] 
+documentation for more information in how these credentials are sourced. 
+
+// TODO: Mention customizing source
```
Comment (enocom) [inline README.md:47]: Maybe something like: See `options.go` for more options to customize the credential source.

File: README.md [hunk 7]
```diff
@@ -1,3 +1,124 @@
...
+
+[adc]: https://cloud.google.com/docs/authentication
+[google-auth]: https://pkg.go.dev/golang.org/x/oauth2/google#hdr-Credentials
+
+### Using the default Dialer
+
+If you don't need to customize your Dialer's behavior, it's convenient to use
```
Comment (enocom) [inline README.md:54]: its

File: README.md [hunk 8]
```diff
@@ -1,3 +1,124 @@
...
+
+// TODO: Mention customizing source
+
+[adc]: https://cloud.google.com/docs/authentication
+[google-auth]: https://pkg.go.dev/golang.org/x/oauth2/google#hdr-Credentials
+
+### Using the default Dialer
+
+If you don't need to customize your Dialer's behavior, it's convenient to use 
+the package's "Dial" option, which initialized a default dialer for you.
```
Comment (enocom) [inline README.md:55]: initializes

File: README.md [hunk 9]
```diff
@@ -1,3 +1,124 @@
...
+  }
+  defer conn.Close(ctx)
+  ```
+[pgconn-cfg]: https://pkg.go.dev/github.com/jackc/pgconn#Config
+</details>
+
+
+### Using DialerOptions
+
+If you need to customize something about the `Dialer`, can and initialize one
```
Comment (enocom) [inline README.md:86]: you can initialize

File: README.md [hunk 10]
```diff
@@ -1,3 +1,124 @@
...
+[adc]: https://cloud.google.com/docs/authentication
+[google-auth]: https://pkg.go.dev/golang.org/x/oauth2/google#hdr-Credentials
+
+### Using the default Dialer
+
+If you don't need to customize your Dialer's behavior, it's convenient to use 
+the package's "Dial" option, which initialized a default dialer for you.  
+<details>
+  <summary>pgx for Postgres</summary>
+   Use the [pgConn.DialFunc field][pgconn-cfg] to create connections:
```
Comment (enocom) [inline README.md:58]: This link doesn't work in the rendered README.

File: README.md [hunk 11]
```diff
@@ -1,3 +1,124 @@
...
+// TODO: Mention customizing source
+
+[adc]: https://cloud.google.com/docs/authentication
+[google-auth]: https://pkg.go.dev/golang.org/x/oauth2/google#hdr-Credentials
+
+### Using the default Dialer
+
+If you don't need to customize your Dialer's behavior, it's convenient to use 
+the package's "Dial" option, which initialized a default dialer for you.  
+<details>
```
Comment (enocom) [inline README.md:56]: Making this a subsection in the README seems fine to me, rather than an expandable section.

**Review by enocom — COMMENTED** [review]
Left some comments.

File: README.md (hunk 2 above)
Comment (kurtisvg) [reply README.md:8]: The database can have both a private and a public IP. I switched to "to an instance's Private IP" for clarity

File: README.md (hunk 5 above)
Comment (kurtisvg) [reply README.md:35]: rephrased.

File: README.md (hunk 7 above)
Comment (kurtisvg) [reply README.md:54]: I think I actually have this one right

File: README.md (hunk 4 above)
Comment (kurtisvg) [reply README.md:22]: I'm okay not explicitly saying this. I think it's probably obvious from the non standard approach.

**Review by enocom — APPROVED** [review]
